### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by tamhv** [lavary/crunz#61]
I used crunz with a php script of sending email ("yiic emailSpool loop" http://cornernote.github.io/yii-email-module). In this script, it'll fetch spooled email (e.g 10 emails) from db and send them, then sleep 1 second, then next 10 emails,...

The problem is the `schedule:run `used very high CPU load

Pls look at process 84672

![screen shot 2017-07-26 at 3 29 11 pm](https://user-images.githubusercontent.com/5440638/28611962-d18f9cbc-7217-11e7-8866-d126bdaf2ae0.png)
![screen shot 2017-07-26 at 3 30 04 pm](https://user-images.githubusercontent.com/5440638/28611983-e2530502-7217-11e7-96f7-7d5447c4e12e.png)

I did try to run `yiic emailSpool loop` without crunz
`#yiic emailSpool loop &`
and it only takes 0.1% cpu (the same result if i put this script to `crontab -e`)
Actually there isn't many emails to send, most of the time this script hit db get 0 spooled emails, sleep, hit db, 0 emails, sleep, ... but the server is always around 90-100% CPU load

**Comment by unrup1** [comment]
Hey, 

the problem is that the runner is not waiting in sleep for your process.
The runner create a extra process for your command.
In the code there is a while loop, which ask every time if the process is closed.

I think because of this loop the cpu usage is so high.

**Comment by tamhv** [comment]
Yeah I'm considering to use supervisord for this kind of script

**Comment by amitt001** [comment]
Any updates on this? @lavary anyone I am also facing 100% cpu usage even when only one job is running. That actual job just takes 1-2% CPU but `php /opt/app/vendor/bin/crunz schedule:run /opt/app/tasks/` takes 100% CPU.

**Comment by adamwinn** [comment]
I am seeing 100% CPU usage too for a simple php task

**Comment by wujekbogdan** [comment]
I'm also experiencing high CPU usage and I'm sure that the usage is not caused by the PHP script I'm running from CRON, but by the crunz script itself.

**Comment by wujekbogdan** [comment]
I did some test over last few days. I found that using `preventOverlapping()` solves the problem of the high CPU usage, but it's not a full solution.

I had several tasks running. All of them were executed in 5 minutes interval (`*/5 * * * *`). All of them had `preventOverlapping()`. I've been monitoring them - each execution was logged, I have also been logging the CPU usage with `sys_getloadavg()`.

The CPU load was low all the time but after several hours one of the tasks stopped executing. So I've uncommented the `preventOverlapping()` in the code. After this change, I've noticed that the task started got back to normal, but I've also noticed that the CPU usage keeps on increasing.

I've contacted server support, and ask support guys to check are there any hanging PHP threads. Of course, there were php processes that consumed CPU.

So for some reason, sometimes PHP processes are not killed after a script is executed. I'm sure that this is not a PHP script itself that consume CPU because if the script were running I would see it in logs. So it must be something related to Crunz scheduler.

**Comment by adamwinn** [comment]
I'd say it's a bug in the Crunz scheduler, but it looks like this project is dead.

**Comment by wujekbogdan** [comment]
@adamwinn 
It's a pity because it's the most powerful CRON manager for PHP and the only competitor for [Jobby](https://github.com/jobbyphp/jobby) library, which is also quite good, but less advanced and also a non-bug-free.

**Comment by kinoute** [comment]
Same problem here. Using `preventOverlapping ()` doesn't fix it unfortunately.

**Comment by kinoute** [comment]
A standard cronjob (like every 5 minutes) linked to a PHP script with a loop inside makes my CPU goes high.

Example (task):

```php
<?php

use Crunz\Schedule;

$schedule = new Schedule();

$schedule->run('/usr/bin/php loop.php')       
         ->everyMinute()
         ->description('Testing a short loop');
         
return $schedule;

```

loop.php (15 secs are enough to see the high CPU usage)

```php
for ($i = 1; $i <= 15; $i++) {
    echo $i;
    sleep(1);
}
```

![capture d ecran 2018-04-12 a 11 47 23](https://user-images.githubusercontent.com/623763/38669843-92de04f2-3e47-11e8-8d49-a468209c135c.png)

Running the loop.php file directly with "php loop.php" doesn't trigger any increase of the CPU.

Link to the code to reproduce the bug: https://github.com/kinoute/crunz-bug

**Comment by PabloKowalczyk** [comment]
Thank you @kinoute, your samples was very helpful. IMO the problem was in the `EventRunner::ManageStartedEvents()`, the `while` loop doesn't have any `(u)sleep`.
Could anyone test CPU usage on this version https://github.com/PabloKowalczyk/crunz/tree/fix-high-cpu-usage? In my tests usage reduced from ~30% to ~0,1% :)

I can do release even today if you confirm that my branch fixed issue.

**Comment by kinoute** [comment]
I tested your fix, now it goes from 96% to 12%. Huge improvement! Even if 12% for an echo in a loop is still a little too much, I will definitely take this fix/release with pleasure!!

**Comment by kinoute** [comment]
Now my code (the one I shared earlier) is using less than 1% of my 2009 Mac Mini CPU! I think it's perfect, thanks!

**Comment by PabloKowalczyk** [comment]
In the end i set probing time to 0,5s, IMO in Cron jobs this time can be set even to 1s.
Issue is fixed and released in v1.5.1. Thanks @kinoute for tests and sample code.
